### PR TITLE
Feat/improve tasks chats

### PR DESCRIPTION
**Pull request by lc-spxl**
- Introduce `PromptTemplate` and use eta to create conditional templates . Overrides can be done by name eg. `<provider>/<model>/<template name>`
- Task introduces a `hint` field at task and field level to be injected in the chat prompt and provide guidance to the LLM on how to handle the task/field state
- Tools use the history to match and infer arguments

### CODE REVIEW

**Review by arghideutis — APPROVED** [review]

File: apps/detection/src/providers/sentiment-analysis/sentiment-analysis.chatgpt.service.ts [hunk 1]
```diff
@@ -31,7 +31,7 @@ Never generate an answer, by default anser with ${defaultEmotion}|1.0`;
     try {
       content = await this.llmProvider.chat({
         system,
-        message,
+        user: message,
```
Comment (Kanthavel-Spindox) [inline apps/detection/src/providers/sentiment-analysis/sentiment-analysis.chatgpt.service.ts:34]: I guess it's just the llmProvider.chat() method signature, but it puzzles me a bit the fact that a user is a message.

File: apps/dialogue/src/dialogue.speech.service.ts [hunk 2]
```diff
@@ -45,19 +45,26 @@ export class DialogueSpeechService {
...
-    if (!targetLanguage) {
-      targetLanguage = await this.session.getLanguage(payload, false);
+    if (!toLanguage) {
+      toLanguage = await this.session.getLanguage(payload, false);
     }
 
-    if (!targetLanguage) return payload.text;
+    if (!toLanguage) return payload.text;
+
+    const fromLanguage = payload.text;
```
Comment (Kanthavel-Spindox) [inline apps/dialogue/src/dialogue.speech.service.ts:58]: Isn't payload.text the message to be translated? Are you using it as a language ISO code?

File: apps/dialogue/src/intent/dialogue.intent.prompt.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+<% } %>
+
+Analyze user interaction in HISTORY and match one of TASKS.
+
+Set the  field 'match' to 'false' in those cases:
+- if there is no match
+- if the assistant already asked for a task in the last two interactions
+- if the user confirmed a task in the last interaction
+
+If the last user message confirm a task proposed by the assistant, set the field "trigger" to true
```
Comment (Kanthavel-Spindox) [inline apps/dialogue/src/intent/dialogue.intent.prompt.ts:32]: "confirms"

File: apps/dialogue/src/intent/dialogue.intent.prompt.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+If the last user message indicate they want to cancel the task, set the field "cancel" to true
```
Comment (Kanthavel-Spindox) [inline apps/dialogue/src/intent/dialogue.intent.prompt.ts:33]: "indicates"

File: apps/dialogue/src/tasks/dialogue.tasks.handler.fields.service.ts [hunk 5]
```diff
@@ -286,7 +278,10 @@ export class DialogueTasksHandlerFieldsService {
       const res = await this.llm.chat<{ result: boolean }>({
         stream: false,
         json: true,
-        message: prompt.join('\n'),
+        user: taskFieldExpressionPrompt({
```
Comment (Kanthavel-Spindox) [inline apps/dialogue/src/tasks/dialogue.tasks.handler.fields.service.ts:281]: The signature with "user" parameter still puzzles me

File: apps/dialogue/src/tasks/dialogue.tasks.handler.fields.prompt.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  value: any;
+  rules: string;
+  language?: string;
+}>(
+  'task-field-validation',
+  `
+Validate and convert the USER value to type <%= data.field.type %> following RULES.
+Provide a reason if the USER value cannot be validated or converted and set value to null
+Answer in parsable JSON format with structure { value: "converted value", reason: "non-technical motivation in case of failure" }
+Do not additional notes or explanation.
```
Comment (Kanthavel-Spindox) [inline apps/dialogue/src/tasks/dialogue.tasks.handler.fields.prompt.ts:47]: "Do not add additional..."

File: apps/dialogue/src/tasks/dialogue.tasks.handler.fields.prompt.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+Given this JSON object:
+<%= data.values %>
+
+<%= data.fieldPrompt %>
+Return a parsable JSON object with format { "result": value }.
+Do not add notes or explanations.`,
+);
+
+export const taskFieldValidationPrompt = PromptTemplate.create<{
+  values: string;
```
Comment (Kanthavel-Spindox) [inline apps/dialogue/src/tasks/dialogue.tasks.handler.fields.prompt.ts:36]: I am not sure I can read this call properly, but two things are bothering me:
- The presence of a "values" parameter (plural) which is not a list or collection
- The presence of both a "value" and a "values" parameters.
- "values" is not used here (but I guess it is used by PromptTemplate.create?)

File: apps/session/src/prompt/session.prompt.service.prompt.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+AVATAR:
+<% if (data.avatar?.name) { %>
+Your name is <%= data.avatar?.name %>. 
+<% } %>
+<% if (data.avatar?.gender) { %>
+Your gender is {data.avatar.gender}.
+<% } %>
+<%= data.avatar?.prompt %>
+<% if (data.emotion) { %>
+Consider the detected user emotion is <%= data.emotion %>, adapt the conversation but do not make it explcit in answer.
```
Comment (Kanthavel-Spindox) [inline apps/session/src/prompt/session.prompt.service.prompt.ts:42]: "explicit"

File: libs/llm/llm.provider.dto.ts [hunk 9]
```diff
@@ -1,50 +1,63 @@
...
   model?: string;
+  tag?: LLMPromptTag;
+  llmCallId?: string;
+};
+
+export type LLMChatRequest = {
   stream?: boolean;
   json?: boolean;
-};
+
```
Comment (Kanthavel-Spindox) [inline libs/llm/llm.provider.dto.ts:36]: extra white line

File: libs/llm/llm.provider.dto.ts [hunk 10]
```diff
@@ -1,50 +1,63 @@
...
 
-export type AvatarChat = LLMChatArgs & {
-  chatArgs?: Partial<LLMChatArgs>;
-  toolsArgs?: Partial<LLMChatArgs>;
-  provider?: LLMProvider;
-  skipChat?: boolean;
-};
+export type AvatarChat = {
+  chat?: PromptTemplateOutput | string;
+
```
Comment (Kanthavel-Spindox) [inline libs/llm/llm.provider.dto.ts:54]: extra white line

File: libs/llm/providers/ollama/adapters/tuda-llama2/tuda.llama2.prompts.ts [hunk 11]
```diff
@@ -1,7 +1,18 @@
-export const llama2Prompt = `Given is the following task-oriented dialog between a human user (<user>) and a virtual agent (<assistant>). 
+import { PromptTemplate } from 'libs/llm/prompt/prompt.template';
+
+export const llama2chatPrompt = PromptTemplate.create(
+  'chat',
+  `
+Given is the following task-oriented dialog between a human user (<user>) and a virtual agent (<assistant>).
```
Comment (Kanthavel-Spindox) [inline libs/llm/providers/ollama/adapters/tuda-llama2/tuda.llama2.prompts.ts:6]: "The following ... is given" ?

File: libs/llm/providers/ollama/adapters/tuda-llama2/tuda.llama2.prompts.ts [hunk 12]
```diff
@@ -1,7 +1,18 @@
-export const llama2Prompt = `Given is the following task-oriented dialog between a human user (<user>) and a virtual agent (<assistant>). 
+import { PromptTemplate } from 'libs/llm/prompt/prompt.template';
+
+export const llama2chatPrompt = PromptTemplate.create(
+  'chat',
+  `
+Given is the following task-oriented dialog between a human user (<user>) and a virtual agent (<assistant>). 
 Previously, this conversation went wrong because the virtual agent made a statement that was contextually incorrect.
 The human user reacted accordingly (No, I don't think so. I asked for assistance with my legal dispute.).
```
Comment (Kanthavel-Spindox) [inline libs/llm/providers/ollama/adapters/tuda-llama2/tuda.llama2.prompts.ts:8]: The example "No, I don't think so. I asked for assistance with my legal dispute." seems very specific. Is it correct?

File: libs/llm/util.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+const logger = new Logger('llm.util');
+
+export const parseJSON = <T = any>(response: string): T | null => {
+  if (!response) return null;
+
+  try {
+    // handle ```json
+    if (response.startsWith('```')) {
+      response = response.substring(3);
```
Comment (Kanthavel-Spindox) [inline libs/llm/util.ts:11]: Can't you just skip the first and the last lines? 
Finding the first and last indexes of '\n' and trimming the response accordingly.

**Review by Kanthavel-Spindox — APPROVED** [review]
Hi,

I'm not sure I can give a relevant help here. 
I just put some comments where the code seemed unclear or where typos are present.

Thanks

File: apps/detection/src/providers/sentiment-analysis/sentiment-analysis.chatgpt.service.ts (hunk 1 above)
Comment (lc-spxl) [reply apps/detection/src/providers/sentiment-analysis/sentiment-analysis.chatgpt.service.ts:34]: yes indeed. The chat accepts messages: LLMMessageDto[] which often is composed of a system message + user message. As a short hand, the API accept a user + message params. API consumer around the code adapts somehow. Do you have a suggestion to improve that snippet ?

File: apps/dialogue/src/dialogue.speech.service.ts (hunk 2 above)
Comment (lc-spxl) [reply apps/dialogue/src/dialogue.speech.service.ts:58]: gotcha, eagle eyes! Thanks!!!

File: apps/dialogue/src/tasks/dialogue.tasks.handler.fields.prompt.ts (hunk 7 above)
Comment (lc-spxl) [reply apps/dialogue/src/tasks/dialogue.tasks.handler.fields.prompt.ts:36]: you are right, values is actually not used in the template and can be removed. thx!

File: libs/llm/providers/ollama/adapters/tuda-llama2/tuda.llama2.prompts.ts (hunk 12 above)
Comment (lc-spxl) [reply libs/llm/providers/ollama/adapters/tuda-llama2/tuda.llama2.prompts.ts:8]: Yes this is a template for a fine-tuned model developed in SERMAS

**Pull request merged**